modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java

```diff
@@ -326,89 +331,113 @@ public synchronized boolean startRaftGroup(
+            nodeOptions.setFsm(new DelegatingStateMachine(lsnr));
 
-        if (groupOptions.snapshotStorageFactory() != null) {
-            serviceFactory.setSnapshotStorageFactory(groupOptions.snapshotStorageFactory());
-        }
+            nodeOptions.setRaftGrpEvtsLsnr(evLsnr);
 
-        if (groupOptions.raftMetaStorageFactory() != null) {
-            serviceFactory.setRaftMetaStorageFactory(groupOptions.raftMetaStorageFactory());
-        }
+            LogStorageFactory logStorageFactory = groupOptions.getLogStorageFactory() == null
+                    ? this.logStorageFactory : groupOptions.getLogStorageFactory();
 
-        nodeOptions.setServiceFactory(serviceFactory);
+            IgniteJraftServiceFactory serviceFactory = new IgniteJraftServiceFactory(logStorageFactory);
 
-        if (initialConf != null) {
-            List<PeerId> mapped = initialConf.stream().map(PeerId::fromPeer).collect(Collectors.toList());
+            if (groupOptions.snapshotStorageFactory() != null) {
+                serviceFactory.setSnapshotStorageFactory(groupOptions.snapshotStorageFactory());
+            }
 
-            nodeOptions.setInitialConf(new Configuration(mapped, null));
-        }
+            if (groupOptions.raftMetaStorageFactory() != null) {
+                serviceFactory.setRaftMetaStorageFactory(groupOptions.raftMetaStorageFactory());
+            }
+
+            nodeOptions.setServiceFactory(serviceFactory);
 
-        IgniteRpcClient client = new IgniteRpcClient(service);
+            if (initialConf != null) {
+                List<PeerId> mapped = initialConf.stream().map(PeerId::fromPeer).collect(Collectors.toList());
 
-        nodeOptions.setRpcClient(client);
+                nodeOptions.setInitialConf(new Configuration(mapped, null));
+            }
+
+            IgniteRpcClient client = new IgniteRpcClient(service);
 
-        NetworkAddress addr = service.topologyService().localMember().address();
+            nodeOptions.setRpcClient(client);
 
-        var peerId = new PeerId(addr.host(), addr.port(), 0, ElectionPriority.DISABLED);
+            NetworkAddress addr = service.topologyService().localMember().address();
 
-        var server = new RaftGroupService(groupId, peerId, nodeOptions, rpcServer, nodeManager);
+            var peerId = new PeerId(addr.host(), addr.port(), 0, ElectionPriority.DISABLED);
 
-        server.start();
+            var server = new RaftGroupService(grpId, peerId, nodeOptions, rpcServer, nodeManager);
 
-        groups.put(groupId, server);
+            server.start();
 
-        return true;
+            groups.put(grpId, server);
+
+            return true;
+        } finally {
+            lock.unlock();
+        }
     }
 
     /** {@inheritDoc} */
     @Override
-    public boolean stopRaftGroup(String groupId) {
-        RaftGroupService svc = groups.remove(groupId);
+    public boolean stopRaftGroup(String grpId) {
```

Comment:
If you call this method twice sequentially, `startGroupInProgressLocks` will contain lock for a stopped group. This may cause OOM, if such groups wont be started anymore.

Reply:
Reworked the whole locking mechanism - now it is simplier and without these issues